task_processing/plugins/kubernetes/kubernetes_pod_executor.py

```diff
@@ -1,61 +1,122 @@
+        # that logic for the threads that operate on them as simple as possible and to make it
+        # possible to cleanly shutdown both of these.
+        # XXX: these can probably be converted to SimpleQueue once all consumers are py37+
+        self.pending_events: "Queue[Union[PodEvent, Sentinel]]" = Queue()
+        self.event_queue: "Queue[Event]" = Queue()
+
+        # TODO(TASKPROC-243): keep track of resourceVersion so that we can continue event processing
+        # from where we left off on restarts
+        self.watch = watch.Watch()
+        self.pod_event_watch_thread = threading.Thread(
+            target=self._pod_event_watch_loop,
+            # ideally this wouldn't be a daemon thread, but a watch.Watch() only checks
+            # if it should stop after receiving an event - and it's possible that we
+            # have periods with no events so instead we'll attempt to stop the watch
+            # and then join() with a small timeout to make sure that, if we shutdown
+            # with the thread alive, we did not drop any events
+            daemon=True,
+        )
+        self.pod_event_watch_thread.start()
+
+        self.pending_event_processing_thread = threading.Thread(
+            target=self._pending_event_processing_loop,
+        )
+        self.pending_event_processing_thread.start()
+
+    def _pod_event_watch_loop(self) -> None:
+        # TODO(TASKPROC-243): we'll need to correctly handle resourceVersion expiration for the case
+        # where the gap between task_proc shutting down and coming back up is long enough for data
+        # to have expired from etcd as well as actually restarting from the last resourceVersion in
+        # case of an exception
+        while not self.stopping:
+            try:
+                for pod_event in self.watch.stream(
+                    self.kube_client.core.list_namespaced_pod,
+                    self.namespace
+                ):
+                    self.pending_events.put(pod_event)
+            except Exception:
+                logger.exception(
+                    "Exception encountered while watching Pod events - restarting watch!")
+
+    def _process_pod_event(self, event: PodEvent) -> None:
+        # XXX: ensure that anything here that updates self.task_metadata acquires a lock
+        if event["type"] == "ADDED":
+            pass
+        elif event["type"] == "DELETED":
+            pass
+        elif event["type"] == "MODIFIED":
+            pass
+        else:
+            logger.warning("Got unknown event with type: {event['type']}")
+
+        # TODO(TASKPROC-245): actually create real events to send to Tron by placing them in
+        # this plugin's event queue
+
+    def _pending_event_processing_loop(self) -> None:
+        event = None
+        while not self.stopping or not self.pending_events.empty():
+            try:
+                event = self.pending_events.get()  # blocks forever until an event is available
```

Comment:
we need to wait with timeout, otherwise it'll never check if `self.stopping` became true

Reply:
my train of thought was that this would be fine since if we're blocking here then there's no pending events and we can let the `join()` timeout in `stop()` and then just shutdown afterwards without fear of losing any data

if we should still add a timeout to this `get()`: we're doing a `join()` with a 1s timeout in `stop()`, so should we use the same timeout here or halve it (i.e, `get(timeout=.5)`)?

Reply:
i see, ok. the problem is that .join may timeout not only because this thread is blocked on .get but also because it takes a long time to process an event. in that case timeout on .join will kill the thread in the middle of something which can lead to corruption.

or are you talking about self.pending_tasks.join? that will actually wait for queue to drop to zero. but then we need to be careful and coordinate shutdown such that no more events are placed on the queue after .join succeeds as this thread will immediately start working on new event and get killed in the middle of processing.

i guess i'd be a bit defensive here and just block on waiting + sleep in between while iterations so that we react to self.stopping reliably.

Reply:
can do!

Reply:
hmm, I guess this approach means that we no longer need to use a poison pill to ensure that we kill this thread since we'll constantly be checking if we should be stopping - i'm gonna go ahead and delete that code then

Reply:
I actually ended up using `pending_events.join()` since that blocks until `task_done()` has been called sufficient times - not until the queue is empty

lemme know your thoughts on https://github.com/Yelp/task_processing/pull/158/commits/cd53f63bad75846b181d455cebccffbf0ff6af76!